train.py

```diff
@@ -302,5 +521,83 @@ def cli_main():
         main(args)
 
 
+def get_args():
+    parser = options.get_training_parser()
+    # TPU: need to control certain flags here.
+    # e.g. parallelization needs to be suppressed and deferred to torch_xla flags
+    # e.g. input tensor shapes need to be controlled via --input_shapes
+    parser.add_argument(
+        '--input_shapes',
+        nargs='*',
+        default=None,
+        help=(
+            'This is used to specify batches and pad lengths. Ex: '
+            '`--input_shapes 256x32 512x16` will produce batches w/ 256 '
+            'sentences padded to length 32, or 512 sentences padded to length '
+            '16. Including too many input shapes will cause graph recompiles and'
+            ' degrade performance. On the other extreme, including 1 shape may '
+            'waste a ton of flops, since batches may contain a lot of pad '
+            'indices on average. Note that the max pad length in this arg will '
+            'be used as `--max-source-positions`'))
+    parser.add_argument('--log_steps', type=int, default=20)
+    parser.add_argument('--num_cores', type=int, default=8)
+    parser.add_argument('--metrics_debug', action='store_true')
+    parser.add_argument('--use_gpu', action='store_true')
+    args = options.parse_args_and_arch(parser)
+    return args
+
+
+def adjust_args_tpu(args):
+    if args.fp16:
+        raise RuntimeError(
+            '--fp16 was provided, this is controlled by env var XLA_USE_BF16')
+    if args.distributed_world_size > 1:
+        xu.eprint('suppressing "distributed_world_size"')
+        args.distributed_world_size = 1
+    if args.distributed_init_method is not None:
+        xu.eprint('suppressing "distributed_init_method"')
+        args.distributed_init_method = None
+    if args.input_shapes is None:
+        raise RuntimeError(
+            'Please specify batches and pad lengths using '
+            '--input_shapes. Ex: `--input_shapes 256x32 512x16` .'
+            'Please refer to the description of the --input_shape'
+            ' arg in --help'
+        )
+    gpu_input_shape_args = ['max_sentences', 'max_sentences_valid', 'max_tokens']
+    nonnull_gpu_input_shape_args = [
+        arg for arg in gpu_input_shape_args if getattr(args, arg) is not None
+    ]
+    if nonnull_gpu_input_shape_args:
+      errmsg = (
+          'On TPUs, please control input shapes '
+          'using `--input_shapes`. Any non-null arg in {} will trigger'
+          ' this error.'
+      ).format(gpu_input_shape_args)
+      raise RuntimeError(errmsg)
+
+    args.input_shapes = parse_input_shapes(args.input_shapes)
+    # XXX (taylanbil): do we ever have more than 2 dimensions in fairseq?
```

Comment:
Just confirming this is still an open question

Reply:
it is.